### DISCUSSION TITLE
[Proposal] Using view block to extend filter blocks

### DISCUSSION THREAD

**Discussion by dinhtungdu** [woocommerce/woocommerce#45154] [WooCommerce, Blocks, and Site Editing]
## The problem

We're working on https://github.com/woocommerce/woocommerce/issues/42255 to create a new set of filter blocks powered by [Interactivity API](https://github.com/WordPress/gutenberg/blob/trunk/packages/interactivity/README.md). The new blocks are server-side rendered, aware of the current product query, always return correct filter counts, and deliver better performance compared to the current ones.

However, there are still some gaps between the new blocks and how online stores are using filters. For example:
- The price filter is a price slider. What if the store wants a price range selector instead (a list of predefined price ranges)?
- Probably the most common filter, how can we display the actual color of the color attribute terms?
- Similar to colors, stores may want to display the image of the attribute terms, like material.

Should WooCommerce ship a color filter block or an image filter block, or both? What about the price range selector?

I don't have the answer to the above questions, but it's clear that **WooCommerce can't cover every possible use case**. But if we can make it possible for developers to customize the output of the filter blocks, I believe we can fill the gap above.

Based on the idea introduced in [#44944](https://github.com/woocommerce/woocommerce/discussions/44944), I want to propose a solution to make our new filter blocks extensible.

## The proposal
I want to propose using view blocks as the extensibility point for new filter blocks.

### What are view blocks?
- They are inner blocks of new filter blocks.
- They consume the filter data provided by parent filter blocks through context.
- They are usually simple blocks to customize the output of filter blocks based on the provided data.
- But nothing stops them from having additional data fetching logic and frontend handler.
- They aren't visible in the inserter and can only be managed through the setting of parent filter blocks.

### How to use view blocks as the extensibility point?
- We need to restructure our new filter blocks to turn them into parent blocks.
- We need to turn view like price slider, checkbox list, dropdown to view blocks.
- We need to provide filter data through the context for inner blocks to consume.
- We need to create and maintain the data and Interactivity API contracts between new filter blocks and view blocks.

### Why should we use view blocks?
- **Allow developers to customize the visual representation of filter blocks**.
- **Improve reusability and prevent duplication**: view blocks can be shared between filter blocks. For example, if we turn Checkbox List and Dropdown into view blocks, they can be shared among Attribute, Rating, and Stock filter blocks.
- **Reduce complexity extending filter blocks**: WooCommerce processes and provides filter data through the block context, both in editors and on the front end. Extenders can consume those contexts to build the visual representation right away without fetching additional data from APIs or building more block UI.
- **Ensure good editing and development experience**: developers can already customize the output of filter block using the `render_block` hook. But it's hard to customize the block using the hook because only raw attributes and rendered content are provided to developers. Besides, in the editor, the block can't be customized (theoretically, they can, but it's super hard to do IMO), this can create a bad experience because the block in the editor looks different from the front end. View blocks give developers control on both the editor and front-end sides.
- **Give developers more freedom**: Although WooCommerce provides some filter data ready to be consumed, we don't limit developers to only using them. They're free to fetch additional data or create complex block UIs within the view block.

### What do we have to do to implement this proposal:

- We have to update all new filter blocks to turn them into parent blocks and split the "view" into dedicated blocks, which are:
  - Price Slider.
  - Checkbox List.
  - Dropdown.
- We need to expose a set of Interactivity API actions meant to be consumed by inner blocks. These actions will become a part of our public APIs.
- We need to expose the necessary data for each filter block type.
- We need documentation and tutorials guiding developers to build view blocks from the foundation we provided.

## What else we've tried

We experimented with some options suggested by @luisherranz in https://github.com/woocommerce/woocommerce/discussions/44944#discussioncomment-7208775:

- Using only one parent block to pass the context down, in this case, the `Product Filter` block: We tried this but turned out the `Product Filter` block became very complex. It's tightly coupled to all inner blocks. Another issue is data flow, at the same time, we need to pass the data from the parent block to the inner as well as pass some data from the inner to the parent. 
- Don't use parent block at all: with this approach, we need to expose some function for the block to get the filter data, but at the same time, developers need to manually detect the product query to be filtered from. 

By the way, in both alternatives we tried, it's quite complex to create the custom filter block. Developers need to repeat almost all logic we create to create block settings, like attribute selectors, and fetching terms from the Store API.

_(See https://github.com/woocommerce/woocommerce/issues/42178#issuecomment-1850737872)_

## Example: Color swatches filter block
How do the reasons above fit with a practical example like color swatches?
- Developers using view block to create the color swatches filter don't have to:
	- Create an attribute selector to select the attribute to filter by.
	- Process filter data to display, like how many products have a `blue` color within a price range between `$10` and `$35` that have 4 or 5 stars rating.
- Developers can start building color swatches right away because all necessary data is available:
	- Selected attribute.
	- Attribute terms of the selected attribute with counts.
- Like normal blocks, developers have complete control of how the color swatches block looks in the editor and how it looks on the front end.
- The color swatches view can have its dedicated settings, like showing the color in a list or grid.
- Developers can decide how the color value is saved and retrieved. It can be a custom term meta, block attribute, or even an external source.

[ Attribute Color Filter - 22 February 2024 - Watch Video](https://www.loom.com/share/5991ffeedd5e4128a621b7c445b2c66c)[![](https://cdn.loom.com/sessions/thumbnails/5991ffeedd5e4128a621b7c445b2c66c-with-play.gif)](https://www.loom.com/share/5991ffeedd5e4128a621b7c445b2c66c)
I created a demo to explain my idea above better:
- I modified the new attribute filter block to:
    - make it a parent block.
    - provide some context to be consumed by the inner block.
- I created a default view block for the checkbox list view.
- I used `@wordpress/create-block` to create the color swatches filter block:
	- It only uses what is available in WordPress Core.
	- The color value is saved to block attributes.
	- I build a custom UI to list the current color for each attribute term and a color picker to let merchants update term color.
- Links:
	- The PR modifying attribute filter: [https://github.com/woocommerce/woocommerce/pull/44888](https://github.com/woocommerce/woocommerce/pull/44888)  
	- The repo of the demo plugin adding color filter block: [https://github.com/dinhtungdu/attribute-color-filter](https://href.li/?https://github.com/dinhtungdu/attribute-color-filter)

## What's next?
What do you think about this? Are there any ideas we should consider before making a decision? Please share your thoughts or feedback below, any feedback is greatly appreciated! Thank you so much for reading!

**Comment by sunyatasattva** [comment]
Thanks for the great proposal @dinhtungdu! The shift towards server-side rendered blocks, marks a significant improvement in performance and accuracy of filter counts, which is a fantastic advancement in itself.

I support the idea of utilizing view blocks as the extensibility point for new filter blocks for the reasons you have already laid out. I am, however, a bit wary for a main reason:

Remember, once we get our feet into something, we can't really easily go back (or at all). I think we need to consider the consequences of this decision very deeply and not just within the domain of the new filters block. Because, in my opinion, if that's the path we take with the Filters, it should be the path we take everywhere else for consistency and DevEx.

   1. So, first question: did you take into consideration how this decision can reverberate in the rest of our work?
   2. I wonder if this will allow us to stay consistent with Gutenberg? I don't want to always slow down our progress as things like this can take forever in Core, but it's actually a big deal.

**Reply by dinhtungdu** [reply]
@sunyatasattva thanks for your support and questions!

> So, first question: did you take into consideration how this decision can reverberate in the rest of our work?

I haven't thought about other parts, but the first thing that pops out in my mind is the `Product Collection` block, which even does a similar thing already. The `Product Collection` block provides `queryId` and `query` through the block context for its inner block to consume. It's the pattern in the Query Loop block as well. From that point of view, every inner block of the `Product Template` can be considered as a view block.

The view block in the proposal above is nothing more than an inner block. We can remove the `view` word eventually to avoid introducing a new term. 

> I wonder if this will allow us to stay consistent with Gutenberg? I don't want to always slow down our progress as things like this can take forever in Core, but it's actually a big deal.

I think my answer to your first question somehow answered this one too. So yes, this will allow us to stay consistent with Gutenberg. We don't use any experimental or planned feature of Gutenberg, we only use inner block - a publicly available API of Gutenberg. We don't have to wait for anything to start implementing the proposal here. Does this answer your question? Looks like I didn't fully get your points.

**Reply by sunyatasattva** [reply]
Sorry if I wasn't clear with my concerns, and thanks for addressing them. Regarding the second concern, what I meant is: will this stay consistent with the direction Gutenberg is taking? What if, in one year, Gutenberg decides to implement their own method of extending “view blocks” (for the Query Loop, or whatever) which clashes with this method we took? Then we'll be stuck on two different tracks and I'd hate to see that.

I just want to make sure this method is future-proof with where Gutenberg will be going, if that makes more sense than what I said in my previous reply.

**Reply by dinhtungdu** [reply]
> will this stay consistent with the direction Gutenberg is taking?

As far as I know and understand, yes, we're only using inner blocks, it's not a new or experimental API. Even if Gutenberg changes from using inner blocks to other things, we can still follow the migration of Query Loop inner blocks and then apply them to our filter blocks. But I don't think it'll happen tbh.

**Reply by samueljseay** [reply]
My main concern about the view block is that we have to expose users to the Woo incubated interactivity API. It's not the stable one released by Gutenberg. I think this concept needs interactivity API to function right now, since we're using it internally. You could argue they can write their own front-end handling but then the purpose of extensibility will be mostly defeated imo.

Even just changing the directives from wc to wp, when the API is formalized will break backwards compatibility unless we promise to keep support for that directive (and any broken, odd or different behaviour of existing directives incubated in Woo). I would be very disappointed to see us be forced to continue incubating our version of interactivity API long after its stable.

It is my personal opinion that there is no rush to offer this extensibility to users until we're using a stable version of interactivity API, because no one will be happy if their use of that API changes in any way. Although what I don't know is how desirable this extensibility really is to developers. 

Extensibility is great, but I think many WordPress devs will not feel confident or ready to use the interactivity API right away, we need to give time for them to adopt and understand that API before they will reach for it to develop customized parts of their ecommerce store, if it looks too confusing, they will just fallback to building their own blocks or custom code. Scaffolding tools could certainly get us some way, but they would have to be *really* good to essentially instruct users how to grok the interactivity API while also extending a block, 2 things that are already a bit daunting I think.

TLDR; What I'm trying to articulate is that, making our interactivity API usage public like this so early, I don't expect we'll see usage, and we'll introduce all the potential problems outlined and risk of backwards support for old or broken behaviours of interactivity API.

There is a bigger question here of education and onboarding for interactivity API, but I think we have to remember just how new this API is within WordPress.

**Reply by dinhtungdu** [reply]
@samueljseay You raised a great point and I totally agree that we shouldn't offer this until the Interactivity API becomes stable, or at least, officially released. At that point, new filter blocks should use the official API, not a customized one like they do right now.

**Reply by sunyatasattva** [reply]
Totally agreed with @samueljseay. Maybe I didn't consider the demo enough, but is that really deeply coupled with the current version of the Interactivity API? I might have understood it incorrectly, but AFAIU, all we are proposing here is that the wrapper blocks will make some data available to the inner blocks.

The inner blocks _should_—of course—be powered by the Interactivity API, but this shouldn't be a mandate, right? There is nothing in the API design that will tie us with a specific implementation of the Interactivity API, apart from the documentation and the examples we'll provide, right? In that case, discussing and even implementing the API design for view blocks doesn't seem blocked to me: maybe publicizing it, and evangelizing it, yes.

Am I misunderstanding here?

**Reply by samueljseay** [reply]
>  I might have understood it incorrectly, but AFAIU, all we are proposing here is that the wrapper blocks will make some data available to the inner blocks.

Well if we just did that alone yes that would be isolated and not tightly coupled to interactivity. @dinhtungdu can correct me if I'm wrong but I think his proposal means to provide a full end-to-end solution and interactivity becomes a key part of it in his design. e.g. he says:

> We need to create and maintain the data and Interactivity API contracts between new filter blocks and view blocks.

In the proposal afaict this is quite key to simplifying the extensibility for developers because for example when they want to (client side) update the query params they'll use our directive to do so. All that said it's not entirely essential. We could leave the front-end part entirely to them, and avoid having examples that use our incubated interactivity (better to show an example with plain JS maybe), but then we better exclude that from this proposal explicitly.

**Reply by dinhtungdu** [reply]
Yep, I want to provide the full end-to-end solution, because apart from visual representation, the block needs to be interactive.

> All that said it's not entirely essential. We could leave the front-end part entirely to them, and avoid having examples that use our incubated interactivity (better to show an example with plain JS maybe), but then we better exclude that from this proposal explicitly.

Can you share more details on this? I'm imaging with this, developers should update the URL manually, but they have to reload the page when filtering right?

**Reply by samueljseay** [reply]
> Can you share more details on this? I'm imaging with this, developers should update the URL manually, but they have to reload the page when filtering right?

Yes @dinhtungdu tbh I hadn't given the alternative a lot of thought. You've confirmed for me that this is a full end-to-end solution you're proposing so thats at least confirmation for me that we need to either:

1. time this closer to maturing the interactivity API 
or
2. somehow limit this proposal to not include interactivity API
